hostapd: disable logging of wpa_passphrase

ckorber opened the pull request:
During the initialization and deinitialization of hostapd, 
the configuration file is logged in syslog and the wpa_passphrase is also logged.

This is not ideal from a security point of view.
Therefore this commit allows the output if compiled in debug mode
(CONFIG_MSG_MIN_PRIORITY <= 2).

Chris1189 commented:
This is example log where wpa_passphrase is logged to syslog:

![Screenshot 2025-01-07 at 15-25-41 root@G3-10002log txt - System6 - TDT Issue Tracking](https://github.com/user-attachments/assets/c3b69af6-ad2b-4051-b79f-6037fac1f215)

In the second to last line `wpa_passphrase` is printed.

ynezz commented on package/network/services/hostapd/patches/804-disable-logging-of-passphrase.patch:
```diff
@@ -0,0 +1,15 @@
+ 		}
+ 	}
+ 
+-	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
```
I get your point, but IMO this should be done differently, masking out the secret(s) only. You're changing behavior, removing complete config dump log line.

Chris1189 replied:
Thank you for pointing that out @KanjiMonster !
So it shouldn't break something if I declare an additional variable `tmp_config_fname`  and copy the contents without wpa_passphrase and log it to syslog, right?
Do you agree @nbd168 ?

guidosarducci replied:
Thanks for trying to tackle this. It has been an unresolved problem since first raised in https://github.com/openwrt/openwrt/issues/14049.

However, I don't agree with just masking out the password. The code dumps far too much information/spam for just the "default" syslog level. Ideally, these details would fall under hostapd's "verbose debug" log level, and log output should respect this. For example, use `MSG_MSGDUMP` instead of `MSG_INFO` for logging.

BTW, are you sure your patch works? I noticed the patch string `"Configuration file: %s (phy %s)%s"` doesn't match what I see in syslog: `"Configuration file: data: ..."`. What am I missing?

Chris1189 replied:
To be honest I didn't look in the details of the mentioned [downstream patch](https://github.com/openwrt/openwrt/blob/main/package/network/services/hostapd/patches/701-reload_config_inline.patch) and I don't want to break behaviour.

Initially my suggestion was to make the printing conditionally. but there were changes requested.

Yes, my patch works. `config_fname`  (changed to a string in mentioned downstream patch) has a certain length and the buffer seems to cut out the `(phy %s)%s` part.
This was my observation during working this out.

ynezz reviewed (changes requested):
We don't want to maintain such downstream changes indefinitely, so please make sure to send this [changes upstream](https://w1.fi/cgit/hostap/tree/CONTRIBUTIONS) and then include the link here, so anyone can find out the upstream status etc.

Commit:
hostapd: disable logging of wpa_passphrase

During the initialization of hostapd, the configuration file is logged in
the syslog and the wpa_passphrase is also logged.

This is not ideal from a security point of view. Therefore this commit allows
the output if compiled in debug mode.

Signed-off-by: Christian Korber <ck@dev.tdt.de>

Chris1189 commented:
> This issue is caused by a downstream patch where we add the support for passing configuration file contents directly instead of a file name: https://github.com/openwrt/openwrt/blob/main/package/network/services/hostapd/patches/701-reload_config_inline.patch.
> 
> So the upstream maintainer is right in that in the upstream version there is no issue, and it just prints a filename.

After @KanjiMonster pointed out that this was caused by a downstream patch, we have two possibilities in my opinion.

1. To not change behaviour of the [701-reload_config_inline.patch](https://github.com/openwrt/openwrt/blob/main/package/network/services/hostapd/patches/701-reload_config_inline.patch) I declared an additional `tmp_config_fname` and cut out the passphrase.

2. Make the logging to syslog conditional (CONFIG_MSG_MIN_PRIORITY <= 2) with my first suggestion.

KanjiMonster commented on package/network/services/hostapd/patches/702-disable-logging-of-passphrase.patch:
```diff
@@ -0,0 +1,69 @@
+ 
++	tmp_config_fname = hostapd_remove_passphrase(config_fname, "wpa_passphrase");
++
+ 	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
+-		   config_fname, phy, iface ? "" : " --> new PHY");
++		   tmp_config_fname, phy, iface ? "" : " --> new PHY");
```
Since the intention of this printf is to print the filename, maybe it would be sufficient to just do something like

```suggestion
 	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
-		   config_fname, phy, iface ? "" : " --> new PHY");
+		   strncmp(name, "data:", 5)) ? config_fname : "<inline>", phy, iface ? "" : " --> new PHY");
```

and skip printing the contents entirely.

KanjiMonster replied:
As context, this would match https://github.com/openwrt/openwrt/blob/main/package/network/services/hostapd/patches/701-reload_config_inline.patch where `name` is set to `<inline>` in this case, so the prints in these functions would also only print that as the filename and not the contents.
